Show content metadata below the description

Stypox opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- [ ] Bugfix (user facing)
- [x] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
This PR adds a metadata menu below the description of video, containing, in order:
- category
- tags (for this @wb9688 suggested to use Chips, but I'd rather not since there could be many tags and it would look bad) (@TiA4f8R after your links PR is merged the tags should be
- licence
- privacy
- age limit (if != NO_AGE_LIMIT)
- language
- support info
- host
- thumbnail url (solves #3724 and #4501 in a different way. Implementing a downloader or a custom image viewer just for a small amount of users does not seem like a good idea to me, but giving access to the thumbnail url at least enables them to view and download the image in the browser)

Each item is hidden if the corresponding value is null or empty.

This PR also disables text selection in the description by default, but adds a button to allow doing that. This fixes the strange flickering observed in #5453, see https://github.com/TeamNewPipe/NewPipe/issues/5453#issuecomment-811363384 for more info.

<details><summary>Screenshot</summary>
<img src="https://user-images.githubusercontent.com/36421898/113312108-3c90a300-930a-11eb-9135-d5a92fbe85e3.png"  width="400px" />
<img src="https://user-images.githubusercontent.com/36421898/113312114-3d293980-930a-11eb-901f-dd8b312411b8.png"  width="400px" />
</details>


#### Fixes the following issue(s)
<!-- Prefix issues with "Fixes" so that GitHub closes them when the PR is merged (note that each "Fixes #" should be in its own item). Also add any other relevant links. -->
- Fixes #3724 
- Fixes #4501 
- Fixes #1502
- Fixes #2991
- Fixes #5453

#### APK testing 
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
On the website the APK can be found by going to the "Checks" tab below the title and then on "artifacts" on the right.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

TobiGr commented on app/src/main/res/values/strings.xml:
```diff
@@ -714,4 +714,17 @@
+    <string name="metadata_language">Language</string>
+    <string name="metadata_support">Support</string>
+    <string name="metadata_host">Host</string>
+    <string name="metadata_thumbnail_url">Thumbnail url</string>
```
```suggestion
    <string name="metadata_thumbnail_url">Thumbnail URL</string>
```

Stypox replied:
This will be made uppercase anyway, it makes no difference, but I will change it

TobiGr commented on app/src/main/java/org/schabi/newpipe/fragments/detail/DescriptionFragment.java:
```diff
@@ -90,4 +97,98 @@ private void setupDescription(final TextView descriptionTextView) {
+            final StringBuilder tags = new StringBuilder();
+            for (int i = 0; i < streamInfo.getTags().size(); ++i) {
+                if (i != 0) {
+                    tags.append(", ");
```
I don't know much about localization, but I think the enumaration comma in Chineese (Mandarin) is `、`.  I don't know about other languages. I am not sure what to do here, because when using YouTube, the tags might be English, but the locale is set to Chineese.

Stypox replied:
Oh, right. Should I create a new translatable string?

mhmdanas replied:
Using chips like @wb9688 suggests would fix this issue, but it might indeed look bad if there are a lot of tags. @Stypox maybe you can give chips a try and post a screenshot so we can see how it would look like? In case they do turn out to look bad, maybe we could show part of them by default and then add some method to show all of them.

Stypox replied:
Ok, I will try

TobiGr reviewed (commented):
Very clean!

opusforlife2 commented:
Killer feature! You're da MVP for the next blog post.

Stypox commented:
@TiA4f8R I fixed #5453 in this PR. See the screenshots and test the APK please ;-)

TiA4f8R commented:
Very good, @Stypox. Just a few suggestions:
- when you click on the button to enable description selection, you should be able to disable text selection. The text when long-pressing the button to enable the description selection should be something like `Enable text description selection` (and `Disable text description selection` if you add the ability to disable the description selection with this button)
- copy metadata to clipboard should be made by long-pressing the tags, the category, etc. of the content, like for comments and may be not by clicking on the category of the metadata item because there is no indication that we can copy to clipboard a metadata item by clicking on this category.

Stypox commented:
> Thumbnail URL @TobiGr 

Done

> you should be able to disable text selection

Done

> copy metadata to clipboard should be made by long-pressing the tags

Done

> maybe you can give chips a try and post a screenshot so we can see how it would look like? @TiA4f8R 

Here are two screenshots. Chips turn out to look good when there are few (left), but I think they use up too much space when there are many (center). Maybe they could be made scrollable (right)? That would make reading them more cumbersome, though.
<img src="https://user-images.githubusercontent.com/36421898/113345726-23e8b300-9333-11eb-860e-a2dbc0f72532.png" width="300px"><img src="https://user-images.githubusercontent.com/36421898/113345729-2519e000-9333-11eb-88ac-603970c0e784.png" width="300px"><img src="https://user-images.githubusercontent.com/36421898/113346097-b426f800-9333-11eb-8252-d615c1c1f71a.png" width="300px">

TiA4f8R commented:
Made tabs scrollable is a good idea and is similar to what the YouTube app does:

![Screenshot_20210401_230448.jpg](https://user-images.githubusercontent.com/74829229/113353929-ae82df80-933e-11eb-8f91-a74d727a0df9.jpg)

I like this option, let's wait for what @opusforlife2 thinks.

Stypox commented:
@TiA4f8R yeah, but that's the case when the tags (i.e. the categories for YouTube app) are sorted by importance, so seeing the first ones is usually enough. But why would you only show the first ~4 tags and hide the other e.g. 15? I am not sure what the best idea is, but I agree the scrollable chip group looks good.

peat80 commented:
Or only show the first x entries to fit in a row and a "more / show all" button that can unfold the rest if there are more? 🤔

Btw. those chips look great and maybe should be used more  through the app to make its look more eye catching... 😃👌

opusforlife2 commented:
Agreed. A 'more' button makes more sense. There could be possible scenarios where the tags take up just enough space that they fit without being cut off, unlike the screenshot. The user would never know that there are more tags that could be scrolled into view.

A 'more' button could expand the tags like in screenshot 1. Scrolling would be cumbersome, like Stypox says.

Plain no-scrolling is definitely bad, because many tags will take up space, and some people tend to tag indiscriminately. Gotta get that sweet, sweet SEO, of course.

SameenAhnaf commented:
> Making tags scrollable is a good idea and is similar to what the YouTube app does:
> 
> ![Screenshot_20210401_230448.jpg](https://user-images.githubusercontent.com/74829229/113353929-ae82df80-933e-11eb-8f91-a74d727a0df9.jpg)
> 
> I like this option, let's wait for what @opusforlife2 thinks.

The same should be added here too on main player.👇
![IMG_20210520_142740](https://user-images.githubusercontent.com/79245641/118946615-61071400-b978-11eb-8839-e512c88e441d.jpg)

TiA4f8R commented:
These tags aren't available because we don't extract them in the extractor, if I am not wrong.

Stypox commented:
I'm (somewhat) back :-D.
I rebased and added the last changes, hopefully:
- use chips for tags instead of a block of text
- clicking on tags opens the search fragment
- long clicking on tags copies to clipboard
- tags are lexicographically sorted, otherwise they would be in random order each time a video is opened (YouTube does not provide them ordered).

See the image and tell me what you think :-) @B0pol @TobiGr 
<img src="https://user-images.githubusercontent.com/36421898/120496955-4db76800-c3be-11eb-9b56-7f20dfa1bfc7.png" width="200px"/>

mhmdanas commented:
@Stypox I tested it and it looks amazing! The only issue I see is the one @opusforlife2 mentioned:
> There could be possible scenarios where the tags take up just enough space that they fit without being cut off, unlike the screenshot. The user would never know that there are more tags that could be scrolled into view.

I don't necessarily find this to be a blocker, but it would be nice if we could find a solution.

mhmdanas commented on app/src/main/java/org/schabi/newpipe/fragments/detail/DescriptionFragment.java:
```diff
@@ -57,37 +71,196 @@ public void onDestroy() {
+            final ItemMetadataTagsBinding itemBinding
+                    = ItemMetadataTagsBinding.inflate(inflater, layout, false);
+
+            Collections.sort(streamInfo.getTags());
```
I'm not confident about this, but maybe the tags should be copied before being mutated?

Stypox replied:
Done

opusforlife2 commented:
Welcome back, @Stypox!

@mhmdanas Does the scroll bar always show if the tags don't fit in the given space? Or does it only become visible upon tapping or scrolling in that area? I'm thinking an easy visual cue is to permanently show the scroll bar for more tags. Then there would be no need for a More button.

mhmdanas commented:
@opusforlife2 no, it only becomes visible on scrolling.

opusforlife2 commented:
Then that's a possible solution. @Stypox thoughts?

opusforlife2 commented:
That thumbnail URL. If it's available like this maybe we could add more functionality on to it. Just the bare URL seems weird to simply show the user.

Maybe an option to Preview, Share or Download it?

TiA4f8R commented:
> Maybe an option to Preview, Share or Download it?

A separate PR would be better for this.

opusforlife2 commented:
Of course. I'm just floating the possibility.

Stypox commented:
> I'm thinking an easy visual cue is to permanently show the scroll bar for more tags

Done. It looks natural, in my opinion

TobiGr reviewed (approved):
LGTM!
Thank you!

**Edit**: tested on tablet and phone emulator with API 30. Android Studio has a bug and I cannot create new emulators currently, so i cannot test this on KitKat, This should be done before merging.

Stypox commented:
@TobiGr I already tested on KitKat, everything seems to work as intended. Sorry, I forgot to mention it.
